Add option to exclude based on paths in XContent

beiske opened the pull request.

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/XContent.java:
```diff
@@ -49,6 +49,11 @@
+     * Creates a new generator using the provided output stream and some
+     * filters.
+     */
+    XContentGenerator createGenerator(OutputStream os, String[] filters, boolean inclusiveFilters) throws IOException;
```
I think that `inclusiveFilters` could be renamed to `inclusive` so that it's consistent with other layers.

Can we also add javadoc here?

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:
```diff
@@ -102,15 +103,28 @@ public XContentBuilder(XContent xContent, OutputStream bos) throws IOException {
+     * inclusiveFilters is false, those matching will be excluded. Make sure to
+     * call {@link #close()} when the builder is done with.
+     */
+    public XContentBuilder(XContent xContent, OutputStream bos, boolean inclusiveFilters, String[] filters) throws IOException {
```
This constructor should be used in `XContentBuilder(XContent xContent, OutputStream bos, String[] filters)` with inclusive set to true

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:
```diff
@@ -158,7 +172,7 @@ public XContentBuilder startObject(String name) throws IOException {
 
     public XContentBuilder startObject(String name, FieldCaseConversion conversion) throws IOException {
-        field(name, conversion);
+        field(name);
```
Why do you need to change this?

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentGenerator.java:
```diff
@@ -121,5 +121,8 @@
+    @Override
     void close() throws IOException;
+
+    void writeStartObject(String name) throws IOException;
```
Why do you need to add this?

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/cbor/CborXContent.java:
```diff
@@ -77,6 +77,11 @@ public XContentGenerator createGenerator(OutputStream os, String[] filters) thro
     }
 
     @Override
+    public XContentGenerator createGenerator(OutputStream os, String[] filters, boolean inclusiveFilters) throws IOException {
```
Same here, `createGenerator(OutputStream os, String[] filters)` could now use this method with inclusive set to true?

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/cbor/CborXContentGenerator.java:
```diff
@@ -20,6 +20,8 @@
 package org.elasticsearch.common.xcontent.cbor;
 
 import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.dataformat.cbor.CBORGenerator;
```
Unused import

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/cbor/CborXContentGenerator.java:
```diff
@@ -34,6 +36,10 @@ public CborXContentGenerator(JsonGenerator jsonGenerator, OutputStream os, Strin
         super(jsonGenerator, os, filters);
     }
 
+    public CborXContentGenerator(JsonGenerator jsonGenerator, OutputStream os, boolean inclusiveFilters, String[] filters) {
```
Same here

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContent.java:
```diff
@@ -81,6 +81,11 @@ public XContentGenerator createGenerator(OutputStream os, String[] filters) thro
     }
 
     @Override
+    public XContentGenerator createGenerator(OutputStream os, String[] filters, boolean inclusiveFilters) throws IOException {
```
Same as above

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilterPathBasedFilter.java:
```diff
@@ -31,25 +31,30 @@
+    private final boolean inclusive;
+
+    public FilterPathBasedFilter(boolean inclusive, FilterPath[] filters) {
+        this.inclusive = inclusive;
```
Can we move this assignation after the `if`?

tlrx commented on core/src/test/java/org/elasticsearch/common/xcontent/support/filtering/FilterPathGeneratorFilteringTests.java:
```diff
@@ -34,56 +34,103 @@
     public void testFilters() throws Exception {
         final String SAMPLE = "{'a':0,'b':true,'c':'c_value','d':[0,1,2],'e':[{'f1':'f1_value','f2':'f2_value'},{'g1':'g1_value','g2':'g2_value'}],'h':{'i':{'j':{'k':{'l':'l_value'}}}}}";
 
-        assertResult(SAMPLE, "a", "{'a':0}");
```
Rename the method to testExclusiveFilters?

tlrx commented on core/src/test/java/org/elasticsearch/common/xcontent/support/filtering/FilterPathGeneratorFilteringTests.java:
```diff
@@ -34,56 +34,103 @@
-        assertResult("{'a':0,'b.c':'value','b':{'c':'c_value'}}", "b.c", "{'b':{'c':'c_value'}}");
-        assertResult("{'a':0,'b.c':'value','b':{'c':'c_value'}}", "b\\.c", "{'b.c':'value'}");
+        assertResult("{'a':0,'b.c':'value','b':{'c':'c_value'}}", true, "b.c", "{'b':{'c':'c_value'}}");
+        assertResult("{'a':0,'b.c':'value','b':{'c':'c_value'}}", true, "b\\.c", "{'b.c':'value'}");
```
Can we add exclusion tests for fields that contain dots?

tlrx commented:
@beiske I left some comments. Have you seen `AbstractFilteringJsonGeneratorTestCase`? I think we need more tests to be complete.

beiske commented:
@tlrx I've implemented most of your suggestions, but in stead of delegating the createGenerator methods per implementation I moved it to a default method in the interface. With this change I think it is also possible to delete the backwards compatible constructors on the content generators.

beiske commented:
Regarding more testing with AbstractFilteringJsonGeneratorTestCase, that probably would not hurt, but I will not have time to address that until after the 26th. So if this PR is relevant for anyone else, it is totally up for grabs. If not I'll continue when I'm back.

beiske commented:
@tlrx I've added more tests. I believe I have addressed all the issues now.

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java:
```diff
@@ -83,6 +83,10 @@ public static XContentBuilder builder(XContent xContent, String[] filters) throw
         return new XContentBuilder(xContent, new BytesStreamOutput(), filters);
     }
 
+    public static XContentBuilder builder(XContent xContent, boolean inclusive, String[] filters) throws IOException {
```
Can the order of arguments be consistent with XContent.createGenerator() method ?

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/smile/SmileXContent.java:
```diff
@@ -68,13 +68,8 @@ public byte streamSeparator() {
-    @Override
-    public XContentGenerator createGenerator(OutputStream os, String[] filters) throws IOException {
-        return new SmileXContentGenerator(smileFactory.createGenerator(os, JsonEncoding.UTF8), os, filters);
+    public XContentGenerator createGenerator(OutputStream os, String[] filters, boolean inclusiveFilters) throws IOException {
```
inclusiveFilters -> inclusive

tlrx commented on core/src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilterPathBasedFilter.java:
```diff
@@ -77,31 +82,39 @@ private TokenFilter evaluate(String name, FilterPath[] filters) {
+            } else if (filter == NO_MATCHING) {
+                return TokenFilter.INCLUDE_ALL;
+            }
         }
```
I'm wondering if this is more readable:
```java
if (include == MATCHING) {
  return inclusive ? TokenFilter.INCLUDE_ALL: null;
}
if (filter == NO_MATCHING) {
  return inclusive ? null : TokenFilter.INCLUDE_ALL;
}

```

beiske replied:
Yeah, I'll agree this time making it more compact makes it easier to see the pattern. I'll change it.

tlrx commented on core/src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractFilteringJsonGeneratorTestCase.java:
```diff
@@ -246,31 +329,207 @@ public void testSimpleFieldWithWildcard() throws Exception {
+                        .endObject()
+                    .endObject()
+                .endObject()
+            .endObject();
```
indent and extra new line

tlrx commented:
Left minor comments

tlrx commented:
LGTM

clintongormley commented:
@beiske this PR is missing documentation (and doesn't even have a description).  Please could you add.

beiske commented:
Added description to PR. This change only affects an internal API and is therefore only documented in javadoc.
